### PR TITLE
Add a way to try private message before final draw

### PR DESCRIPTION
**Pull request by pyrech**
capture du 2018-10-29 14-31-22

### CODE REVIEW

**Comment by damienalexandre** [comment]
I think the BIG button on top of the real button is confusing, the "sample" message button should be smaller and maybe not here. What do you think about putting it as a small button just after the textarea?

**Comment by pyrech** [comment]
Completely agree, will make the change later :wink:

**Comment by pyrech** [comment]
Latest version

capture du 2018-10-30 14-17-23

File: src/Controller/SantaController.php [hunk 1]
```diff
@@ -119,6 +122,55 @@ public function run(MessageDispatcher $messageDispatcher, Rudolph $rudolph, Requ
         return new Response($content);
     }
 
+    public function sendSampleMessage(Request $request, string $application): Response
+    {
+        if (!$request->isMethod('POST') || !$request->isXmlHttpRequest()) {
```
Comment (lyrixx) [inline src/Controller/SantaController.php]: This could be done by the routing

File: src/Controller/SantaController.php [hunk 2]
```diff
@@ -119,6 +122,55 @@ public function run(MessageDispatcher $messageDispatcher, Rudolph $rudolph, Requ
         return new Response($content);
     }
 
+    public function sendSampleMessage(Request $request, string $application): Response
```
Comment (lyrixx) [inline src/Controller/SantaController.php]: applicationName

File: src/Controller/SantaController.php [hunk 3]
```diff
@@ -119,6 +122,55 @@ public function run(MessageDispatcher $messageDispatcher, Rudolph $rudolph, Requ
...
+    {
+        if (!$request->isMethod('POST') || !$request->isXmlHttpRequest()) {
+            throw $this->createNotFoundException();
+        }
+
+        $application = $this->getApplication($application);
+        // An admin is required to use the sample feature
+        if (!$application->isAuthenticated() || !$application->getAdmin()) {
```
Comment (lyrixx) [inline src/Controller/SantaController.php]: No voter ?

File: src/Controller/SantaController.php [hunk 4]
```diff
@@ -119,6 +122,55 @@ public function run(MessageDispatcher $messageDispatcher, Rudolph $rudolph, Requ
...
+            throw $this->createAccessDeniedException();
+        }
+
+        $message = $request->request->get('message');
+
+        $errors = $this->validate([], $message, true);
+
+        if (\count($errors) < 1) {
```
Comment (lyrixx) [inline src/Controller/SantaController.php]: I would handle the case where there are errors in the if.

File: src/Controller/SantaController.php [hunk 5]
```diff
@@ -119,6 +122,55 @@ public function run(MessageDispatcher $messageDispatcher, Rudolph $rudolph, Requ
...
+
+            return new JsonResponse([
+                'success' => true,
+            ]);
+        }
+
+        return new JsonResponse([
+            'success' => false,
+            'errors' => $errors,
+        ]);
```
Comment (lyrixx) [inline src/Controller/SantaController.php]: Proper status code is better

File: src/Controller/SantaController.php [hunk 6]
```diff
@@ -119,6 +122,55 @@ public function run(MessageDispatcher $messageDispatcher, Rudolph $rudolph, Requ
...
+
+            try {
+                $application->sendSecretMessage($secretSanta, $application->getAdmin()->getIdentifier(), $application->getAdmin()->getIdentifier(), true);
+            } catch (MessageSendFailedException $e) {
+                return new JsonResponse([
+                    'success' => false,
+                    'errors' => [
+                        'send' => 'Error when sending the sample message',
+                    ],
+                ]);
```
Comment (lyrixx) [inline src/Controller/SantaController.php]: Proper status code is better

File: src/Controller/SantaController.php (hunk 2 above)
Comment (pyrech) [reply src/Controller/SantaController.php]: No, it's the name of the route parameter

File: src/Controller/SantaController.php (hunk 3 above)
Comment (pyrech) [reply src/Controller/SantaController.php]: No security in this projet

File: src/Controller/SantaController.php (hunk 5 above)
Comment (pyrech) [reply src/Controller/SantaController.php]: Js part don't care about the error code, so not sure it's worth it

**Pull request merged**